JBMC: Added java-models-library dependency

Degiorgio opened the pull request:
JBMC: Added java-models-library dependency

This commit adds a dependency to the java-models-library
(https://github.com/diffblue/java-models-library). This repository
contains models for a number of classes derived from the java standard
library. These models are needed, i.e., to support concurrency.

This means that the process of building JBMC has changed slightly as
one first needs to download the java-models-library:

```bash
  make -C jbmc/src java-models-library-download
  make -C jbmc/src
```
Due possible licensing issues, the ability to automatically embed the
java core models into JBMC has been removed. Instead, one must
explicitly use the '--classpath' option to load the models.
Consequently, the '--no-core-models' option and related code was
removed as it is no longer relevant.

Commit also adds a new make target, `make dist`. This target in
addition to building jbmc will create a `dist/` directory with two
sub-folders, `bin` and `lib`. Executables will be copied to the former,
while 'core-models.jar' will copied to the latter.

Note: `src/org/cprover/CProver.java` has also been removed as this has
been superseded by the CProver.java in the java-models-library.

TEST-GEN-POINTER-BUMP: https://github.com/diffblue/test-gen/pull/1878

chrisr-diffblue commented on jbmc/src/Makefile:
```diff
@@ -48,3 +48,13 @@ $(patsubst %, %_clean, $(DIRS)):
+
+java-models-library-download:
+	@echo "Downloading java models library"
+	@curl -L https://github.com/diffblue/java-models-library/archive/master.zip > java-models-library.tar.gz
```
This seems wrong - you are downloading a .zip file, but saving it with a .tar.gz extension...

Degiorgio replied:
oops, it still works but its misleading, fixed.

chrisr-diffblue commented on jbmc/src/java_bytecode/library/converter.cpp:
```diff
@@ -66,7 +66,7 @@ int main(int argc, char *argv[])
   }
 
-  std::cout << "\n#define " << varname <<  "_SIZE " << size << "\n\n";
+  std::cout << "\n#define " << varname <<  "_SIZE " << size << "\n";
```
I know you mentioned this change in the PR description, and I don't have strong feels whether it belongs in the PR or not, but if it's in this PR it should probably be in a commit of it's own.

Degiorgio replied:
done

tautschnig commented on .travis.yml:
```diff
@@ -286,4 +288,4 @@ notifications:
     on_start: never
     on_cancel: never
     on_error: always
-    
```
Nit pick: I think this blank line should be removed completely.

Degiorgio replied:
done

tautschnig commented on jbmc/src/java_bytecode/library/.gitignore:
```diff
@@ -0,0 +1,5 @@
+classes
+converter
+core-models.jar
+java_core_models.inc
```
We don't do per-directory gitignore files elsewhere, so this might cause confusion. I'd suggest to include this in the top-level .gitignore file.

Degiorgio replied:
done

Degiorgio committed:
JBMC: Added java-models-library dependency

This commit adds a dependency to the java-models-library
(https://github.com/diffblue/java-models-library). This repository
contains models for number of classes derived from the java standard
library. These models are needed to support concurrency.

This means that the process of building JBMC has changed slightly as
one first needs to download the java-models-library. I.E:

  make -C jbmc/src java-models-library-download
  make -C jbmc/src

Due possible licensing issues, the ability to automatically embed the
java core models into JBMC has been removed. Instead, one must
explicitly use the '--classpath' option to load the models.
Consequently, the '--no-core-models' option and related code was
removed as it is no longer relevant.

Commit also adds a new make target, 'make dist'. This target in
addition to building jbmc will create a 'dist' directory with two
sub-folders, bin and lib. Executables will be copied to the former,
while 'core-models.jar' will copied to the latter.

Note: src/org/cprover/CProver.java has also been removed as this has
been superseded by the CProver.java in the java-models-library.

tautschnig reviewed (commented):
If I understand correctly, the jbmc binary will include the model library (as a large string). Does the Classpath Exception make the resulting jbmc binary distributable under the BSD license? My reading is that that's the case, but more knowledgeable people's input would be good to have.

Degiorgio commented:
@tautschnig yes this is correct, @peterschrammel ?

peterschrammel reviewed (commented):
With the risk of kicking off a side-discussion: Is there a way for the user to obtain a copy of the `org.cprover` package as a jar file? The user will need it in order to *compile* any java code that directly uses `org.cprover` methods. Maybe `org.cprover` should be published as a separate package on maven?

tautschnig commented:
Knowing nothing about maven: cbmc and goto-cc have dist-* scripts that put together archives that eventually go on the website. I'm not sure what the intended distribution story for jbmc is, but it may be worth adding such scripts that would then also ship the compiled jar file?

Degiorgio commented:
@peterschrammel, when jbmc/src/java_bytecode/library is built  core-models.jar gets generated. This contains both the **org.cprover** and java model classes. If we need a jar file  with only org.cprover, then we would need to  modify the build scripts (either in CBMC or the java-models-library repository)

peterschrammel commented:
@Degiorgio, can you put core-models.jar on the classpath without getting conflicts with rt.jar?

Degiorgio commented:
@peterschrammel , as discussed, I will modify the build scripts to generate two jars files:

1:  core-models.jar: containing only java models
2:  cprover.jar:  containing only org.cprover 

will only embed  core-models.jar into jbmc executable.

smowton commented:
Does this mean jbmc will *always* run with the models lib, regardless of the input? I ask as it's pretty heavy, which will prevent any non-lazy-loading test from running.

Degiorgio commented:
@smowton , by default yes, but you can disable the models by specifying the '--no-core-models'  flag.  Will run some tests on this. Worst case we will make it opt-in instead of opt-out

Degiorgio committed:
JBMC: Minor fix, removing superfluous padding from converter

peterschrammel commented:
We should keep it 'opt out' because no 'normal user' would be able to run the tool without core models.
I have commits in https://github.com/peterschrammel/cbmc/tree/cav18-rebased that clean up all the tests to enable lazy loading where necessary, but requires further rebasing now.

Degiorgio commented:
@peterschrammel out of abundance of caution, would it be acceptable, to make it 'opt-in' at first? My reasoning  here is that without the lazy loading option it will lead to a blow-up, even for the simplest of use-cases.  I think if we where to make it it 'opt-out' we should also make lazy loading 'opt-out'.

Degiorgio commented:
@peterschrammel @tautschnig  updated to address possible licensing issues.

Degiorgio commented:
@peterschrammel I will split core-models.jar (into two jars, one for core-models and one for CProver  API) in a separate pr.

cesaro commented on jbmc/README.md:
````diff
@@ -30,8 +30,10 @@ Compilation
+make -C jbmc/src java-models-library-download
 make -C jbmc/src
 ```
+
````
The file `COMPILING.md` should be updated as well.

Degiorgio replied:
done

cesaro commented on jbmc/src/Makefile:
```diff
@@ -1,4 +1,5 @@
 DIRS = janalyzer jbmc jdiff java_bytecode miniz
+ROOT = ../
```
Is this variable necessary? You could obtain the path pointed by it via `$(CPROVER_DIR)`?

Degiorgio replied:
CPROVER_DIR gives the path to the project root, since the generated files are for JBMC, it kind does not make sense for the JBMC distribution folder to be in the root of the repository (it should be in $(CPROVER_DIR)/jbmc

cesaro replied:
That's what I meant, you could use `$(CPROVER_DIR)/jbmc` wherever you use `ROOT`. You can ignore this, it's a small thing.

cesaro commented on jbmc/src/Makefile:
```diff
@@ -40,11 +41,33 @@ generated_files: $(patsubst %, %_generated_files, $(DIRS))
+# extended JBMC models download, for your convenience
+java-models-library-download:
+	@echo "Downloading java models library"
+	@curl -L https://github.com/diffblue/java-models-library/archive/master.zip > java-models-library.zip
```
I would have use the `/tmp` folder as a temporary folder to store the zip file, but that's not very important.

Degiorgio replied:
I took this approach as its the one being used for minisat, for the sake of consistency, will leave it as is.

peterschrammel replied:
Could you use `wget` here (as we do for minisat-download)?

tautschnig replied:
The minisat download uses (by default) `lwp-download`, which doesn't pick up the OS certificate chain required for HTTPS. I'm tempted to suggest to get rid of `lwp-download` instead.

cesaro commented on jbmc/src/Makefile:
```diff
@@ -40,11 +41,33 @@ generated_files: $(patsubst %, %_generated_files, $(DIRS))
+
+.PHONY: dist
+dist: java-models-library-download all
+	@mkdir --p $(ROOT)dist/lib
```
`-p` instead of `--p`.

Degiorgio replied:
done

cesaro commented on jbmc/src/Makefile:
```diff
@@ -40,11 +41,33 @@ generated_files: $(patsubst %, %_generated_files, $(DIRS))
+	@curl -L https://github.com/diffblue/java-models-library/archive/master.zip > java-models-library.zip
+	@unzip java-models-library.zip
+	@rm java-models-library.zip
+	@cp -r java-models-library-master/src java_bytecode/library
```
I see no reason to hide the commands here with `@`. I would remove it.

Degiorgio replied:
done

cesaro commented on jbmc/src/Makefile:
```diff
@@ -40,11 +41,33 @@ generated_files: $(patsubst %, %_generated_files, $(DIRS))
+	@cp java_bytecode/library/core-models.jar $(ROOT)dist/lib
+	@mkdir --p $(ROOT)dist/bin
+	@cp jbmc/jbmc $(ROOT)dist/bin
+	@cp janalyzer/janalyzer $(ROOT)dist/bin
```
I would definitively like to see what `make dist` is doing when I type it, better to remove the `@`.

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/library/CMakeLists.txt:
```diff
@@ -1,5 +1,14 @@
+include("${CBMC_SOURCE_DIR}/../cmake/DownloadProject.cmake")
+download_project(PROJ java_models_library
+    URL https://github.com/diffblue/java-models-library/archive/master.zip
+    PATCH_COMMAND cmake -E copy_directory "${JBMC_BINARY_DIR}/../../java_models_library-src/src"
```
Please document why you use the `PATCH_COMMAND` ;)

Degiorgio replied:
done

cesaro commented on jbmc/src/java_bytecode/library/CMakeLists.txt:
```diff
@@ -9,16 +18,10 @@ add_executable(java-converter converter.cpp)
-    DEPENDS ${JAVA_CORE_MODELS_INC})
+# copy 'core-models.jar' to '<PROJECT_ROOT>/jbmc/src/java_bytecode/library'.
+# This is needed to deal with unit tests that make use of the core-models
+# library.
```
add `so that they can find the 'core-models.jar' in the same place as if the project had been compiled with 'make'.`

Degiorgio replied:
done

cesaro reviewed (changes requested):
A couple of things to change

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_language.h:
```diff
@@ -40,7 +39,6 @@ Author: Daniel Kroening, kroening@kroening.com
   "(java-no-load-class):"
 
 #define JAVA_BYTECODE_LANGUAGE_OPTIONS_HELP /*NOLINT*/                                          \
-  " --no-core-models                 don't load internally provided models for core classes in\n"/* NOLINT(*) */ \
```
You have to remove the next line too!

Degiorgio replied:
oops,fixed.

peterschrammel commented on .gitignore:
```diff
@@ -122,9 +121,13 @@ src/ansi-c/file_converter
 src/ansi-c/library/converter.exe
-jbmc/src/java_bytecode/converter
-jbmc/src/java_bytecode/converter.exe
+jbmc/src/java_bytecode/library/converter.exe
```
This should probably go into the PR that is on hold.

peterschrammel commented on jbmc/src/java_bytecode/library/converter.cpp:
```diff
@@ -66,7 +66,7 @@ int main(int argc, char *argv[])
   }
 
-  std::cout << "\n#define " << varname <<  "_SIZE " << size << "\n\n";
+  std::cout << "\n#define " << varname << "_SIZE " << size << "\n";
```
This file should be moved to the other PR.

cesaro replied:
@peterschrammel I'm not sure I get this. The converter program was committed long time ago. This just fixes a bug in it, I see no value in moving this bug fix to another PR. It has to be fixed no matter what?

peterschrammel replied:
ok

Degiorgio commented:
@peterschrammel  addressed the issues you raised, with regards to the minor fix, my opinion is that it should  stay in the PR as it is a bug that is currently present in develop and  it is in a way related to  the PR (although admittedly  a bit contradictory, as in one commit we are removing core-models and in the other-one we are fixing it).  Having said this, I have no issues with dropping  it from this PR.
